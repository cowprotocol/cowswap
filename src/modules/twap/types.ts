import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { SafeTransactionParams } from '../../common/types'

// Read more: https://github.com/rndlabs/composable-cow#data-structure
export interface TWAPOrder {
  sellAmount: CurrencyAmount<Token>
  buyAmount: CurrencyAmount<Token>
  receiver: string
  numOfParts: number
  startTime: number
  timeInterval: number
  span: number
  appData: string
}

export interface TWAPOrderStruct {
  sellToken: string
  buyToken: string
  receiver: string
  partSellAmount: string
  minPartLimit: string
  // timeStart
  t0: number
  // numOfParts
  n: number
  // timeInterval
  t: number
  span: number
  appData: string
}

export enum TwapOrderStatus {
  WaitSigning = 'WaitSigning',
  Pending = 'Pending',
  Scheduled = 'Scheduled',
  Cancelled = 'Cancelled',
  Cancelling = 'Cancelling',
  Expired = 'Expired',
  Fulfilled = 'Fulfilled',
}

export interface TwapOrdersSafeData {
  conditionalOrderParams: ConditionalOrderParams
  safeTxParams: SafeTransactionParams
}

export interface TwapOrderExecutionInfo {
  executedSellAmount: string
  executedBuyAmount: string
  executedFeeAmount: string
}

export interface TwapOrderItem {
  order: TWAPOrderStruct
  status: TwapOrderStatus
  chainId: SupportedChainId
  executedDate?: string
  submissionDate: string
  safeAddress: string
  id: string
  safeTxParams?: SafeTransactionParams
  executionInfo: TwapOrderExecutionInfo
}

export interface ConditionalOrderParams {
  staticInput: string
  salt: string
  handler: string
}

export interface TwapOrderInfo {
  id: string
  orderStruct: TWAPOrderStruct
  safeData: TwapOrdersSafeData
}

export type TwapOrdersAuthResult = { [key: string]: boolean | undefined }
