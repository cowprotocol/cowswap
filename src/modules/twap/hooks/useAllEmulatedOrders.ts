import { useMemo } from 'react'

import { OrderClass } from '@cowprotocol/cow-sdk'

import { Order } from 'legacy/state/orders/actions'
import { useOrders } from 'legacy/state/orders/hooks'

import { useIsSafeApp, useWalletInfo } from 'modules/wallet'

import { useEmulatedPartOrders } from './useEmulatedPartOrders'
import { useEmulatedTwapOrders } from './useEmulatedTwapOrders'

export function useAllEmulatedOrders(): Order[] {
  const { chainId, account } = useWalletInfo()
  const emulatedTwapOrders = useEmulatedTwapOrders()
  const emulatedPartOrders = useEmulatedPartOrders()
  const isSafeApp = useIsSafeApp()

  const limitOrders = useOrders(chainId, account, OrderClass.LIMIT)
  const discreteTwapOrders = useMemo(() => {
    return limitOrders.filter((order) => order.composableCowInfo?.isVirtualPart === false)
  }, [limitOrders])

  const allEmulatedOrders = useMemo(() => {
    if (!isSafeApp) return []

    return emulatedTwapOrders.concat(emulatedPartOrders).concat(discreteTwapOrders)
  }, [emulatedTwapOrders, emulatedPartOrders, discreteTwapOrders, isSafeApp])

  return allEmulatedOrders
}
