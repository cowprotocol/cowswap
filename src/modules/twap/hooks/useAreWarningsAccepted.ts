import { useAtomValue } from 'jotai'

import { useIsFallbackHandlerRequired } from './useFallbackHandlerVerification'
import { useTwapWarningsContext } from './useTwapWarningsContext'

import { twapOrdersSettingsAtom } from '../state/twapOrdersSettingsAtom'

export function useAreWarningsAccepted(): boolean {
  const { isPriceImpactAccepted, isFallbackHandlerSetupAccepted } = useAtomValue(twapOrdersSettingsAtom)
  const { showPriceImpactWarning } = useTwapWarningsContext()
  const isFallbackHandlerRequired = useIsFallbackHandlerRequired()

  if (showPriceImpactWarning) return isPriceImpactAccepted

  if (isFallbackHandlerRequired) return isFallbackHandlerSetupAccepted

  return true
}
