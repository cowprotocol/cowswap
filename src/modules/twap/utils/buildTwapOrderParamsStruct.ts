import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { defaultAbiCoder } from '@ethersproject/abi'
import { hexZeroPad } from '@ethersproject/bytes'

import { twapOrderToStruct } from './twapOrderToStruct'

import { TWAP_HANDLER_ADDRESS, TWAP_ORDER_STRUCT } from '../const'
import { ConditionalOrderParams, TWAPOrder } from '../types'

// TODO: support other conditional orders (stop loss, GAT, etc.)
export function buildTwapOrderParamsStruct(chainId: SupportedChainId, order: TWAPOrder): ConditionalOrderParams {
  const twapOrderData = twapOrderToStruct(order)

  return {
    handler: TWAP_HANDLER_ADDRESS[chainId],
    salt: hexZeroPad(Buffer.from(Date.now().toString(16), 'hex'), 32),
    staticInput: defaultAbiCoder.encode([TWAP_ORDER_STRUCT], [twapOrderData]),
  }
}
