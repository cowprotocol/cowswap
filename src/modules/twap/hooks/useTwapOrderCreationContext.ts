import { Erc20 } from '@cowprotocol/abis'
import { ComposableCoW } from '@cowprotocol/abis'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { Nullish } from 'types'

import { useTokenContract } from 'legacy/hooks/useContract'

import { CURRENT_BLOCK_FACTORY_ADDRESS } from 'modules/advancedOrders'
import { useComposableCowContract } from 'modules/advancedOrders/hooks/useComposableCowContract'
import { useWalletInfo } from 'modules/wallet'

import { useNeedsApproval } from 'common/hooks/useNeedsApproval'
import { useTradeSpenderAddress } from 'common/hooks/useTradeSpenderAddress'

export interface TwapOrderCreationContext {
  composableCowContract: ComposableCoW
  needsApproval: boolean
  spender: string
  currentBlockFactoryAddress: string
  erc20Contract: Erc20
}

export function useTwapOrderCreationContext(
  inputAmount: Nullish<CurrencyAmount<Token>>
): TwapOrderCreationContext | null {
  const { chainId } = useWalletInfo()
  const composableCowContract = useComposableCowContract()
  const needsApproval = useNeedsApproval(inputAmount)
  const erc20Contract = useTokenContract(inputAmount?.currency.address)
  const spender = useTradeSpenderAddress()
  const currentBlockFactoryAddress = chainId ? CURRENT_BLOCK_FACTORY_ADDRESS[chainId] : null

  if (!composableCowContract || !erc20Contract || !spender || !currentBlockFactoryAddress) return null

  return { composableCowContract, erc20Contract, needsApproval, spender, currentBlockFactoryAddress }
}
