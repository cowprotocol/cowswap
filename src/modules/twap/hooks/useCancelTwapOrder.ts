import { useAtomValue } from 'jotai/utils'
import { useCallback } from 'react'

import { useGP2SettlementContract } from 'legacy/hooks/useContract'
import { Order } from 'legacy/state/orders/actions'

import { useComposableCowContract } from 'modules/advancedOrders/hooks/useComposableCowContract'
import { useSafeAppsSdk } from 'modules/wallet/web3-react/hooks/useSafeAppsSdk'

import type { OnChainCancellation } from 'common/hooks/useCancelOrder/onChainCancellation'

import { cancelTwapOrderTxs, estimateCancelTwapOrderTxs } from '../services/cancelTwapOrderTxs'
import { twapPartOrdersListAtom } from '../state/twapPartOrdersAtom'

export function useCancelTwapOrder(): (order: Order) => Promise<OnChainCancellation> {
  const twapPartOrders = useAtomValue(twapPartOrdersListAtom)
  const safeAppsSdk = useSafeAppsSdk()
  const settlementContract = useGP2SettlementContract()
  const composableCowContract = useComposableCowContract()

  return useCallback(
    async (order: Order) => {
      const orderId = order.composableCowInfo?.id

      if (!orderId) {
        throw new Error('Wrong orderId for TWAP order cancellation')
      }

      if (!composableCowContract || !settlementContract || !safeAppsSdk) {
        throw new Error('Context is not full to cancel TWAP order')
      }

      const partOrder = twapPartOrders.find((item) => item.twapOrderId === orderId)
      const partOrderId = partOrder?.uid

      const context = { composableCowContract, settlementContract, orderId, partOrderId }

      return {
        estimatedGas: await estimateCancelTwapOrderTxs(context),
        sendTransaction: (processCancelledOrder) => {
          return safeAppsSdk.txs.send({ txs: cancelTwapOrderTxs(context) }).then((res) => {
            const txHash = res.safeTxHash
            const sellTokenAddress = order.inputToken.address
            const sellTokenSymbol = order.inputToken.symbol

            processCancelledOrder({ txHash, orderId, sellTokenAddress, sellTokenSymbol })

            if (partOrderId) {
              processCancelledOrder({ txHash, orderId: partOrderId, sellTokenAddress, sellTokenSymbol })
            }
          })
        },
      }
    },
    [composableCowContract, settlementContract, safeAppsSdk, twapPartOrders]
  )
}
