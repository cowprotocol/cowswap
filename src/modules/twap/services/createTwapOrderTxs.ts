import { IConditionalOrder } from 'abis/types/ComposableCoW'
import { defaultAbiCoder } from '@ethersproject/abi'
import { MetaTransactionData } from '@safe-global/safe-core-sdk-types'
import { TWAPOrder, TWAPOrderStruct } from '../types'
import { TWAP_HANDLER_ADDRESS, TWAP_ORDER_STRUCT } from '../const'
import { COMPOSABLE_COW_ADDRESS } from 'modules/advancedOrders'
import { hexZeroPad } from '@ethersproject/bytes'
import { TwapOrderCreationContext } from '../hooks/useTwapOrderCreationContext'
import { SupportedChainId } from '@cowprotocol/cow-sdk'

function getTwapOrderParamsStruct(
  chainId: SupportedChainId,
  order: TWAPOrder
): IConditionalOrder.ConditionalOrderParamsStruct {
  const twapOrderData: TWAPOrderStruct = {
    sellToken: order.sellAmount.currency.address,
    buyToken: order.buyAmount.currency.address,
    receiver: order.receiver,
    partSellAmount: order.sellAmount.divide(order.numOfParts).quotient.toString(),
    minPartLimit: order.buyAmount.divide(order.numOfParts).quotient.toString(),
    t0: order.startTime,
    n: order.numOfParts,
    t: order.timeInterval,
    span: order.span,
  }

  return {
    handler: TWAP_HANDLER_ADDRESS[chainId],
    salt: hexZeroPad(Buffer.from(Date.now().toString(16), 'hex'), 32),
    staticInput: defaultAbiCoder.encode([TWAP_ORDER_STRUCT], [twapOrderData]),
  }
}

export function createTwapOrderTxs(order: TWAPOrder, context: TwapOrderCreationContext): MetaTransactionData[] {
  const { chainId, composableCowContract, needsApproval, erc20Contract, spender } = context

  const sellTokenAddress = order.sellAmount.currency.address
  const sellAmountAtoms = order.sellAmount.quotient.toString()

  // TODO: support other conditional orders (stop loss, GAT, etc.)
  const creationParams = getTwapOrderParamsStruct(chainId, order)

  const createOrderTx = {
    to: COMPOSABLE_COW_ADDRESS[chainId],
    data: composableCowContract.interface.encodeFunctionData('create', [creationParams, true]),
    value: '0',
    operation: 0,
  }

  if (!needsApproval) return [createOrderTx]

  const approveTx = {
    to: sellTokenAddress,
    data: erc20Contract.interface.encodeFunctionData('approve', [spender, sellAmountAtoms]),
    value: '0',
    operation: 0,
  }

  return [approveTx, createOrderTx]
}
