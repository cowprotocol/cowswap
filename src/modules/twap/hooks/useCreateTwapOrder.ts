import { useAtomValue } from 'jotai'
import { useUpdateAtom } from 'jotai/utils'
import { useCallback } from 'react'

import { OrderClass, OrderKind } from '@cowprotocol/cow-sdk'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { Nullish } from 'types'

import { twapConversionAnalytics } from 'legacy/components/analytics/events/twapEvents'
import store from 'legacy/state'
import { dispatchPresignedOrderPosted } from 'legacy/state/orders/middleware/updateOrderPopup'
import { getCowSoundSend } from 'legacy/utils/sound'
import { getOrderSubmitSummary } from 'legacy/utils/trade'

import { useAdvancedOrdersDerivedState } from 'modules/advancedOrders'
import { useAppData, useUploadAppData } from 'modules/appData'
import { useTradeConfirmActions, useTradePriceImpact } from 'modules/trade'
import { SwapFlowAnalyticsContext, tradeFlowAnalytics } from 'modules/trade/utils/analytics'
import { useWalletInfo } from 'modules/wallet'
import { useSafeAppsSdk } from 'modules/wallet/web3-react/hooks/useSafeAppsSdk'

import { useConfirmPriceImpactWithoutFee } from 'common/hooks/useConfirmPriceImpactWithoutFee'

import { useExtensibleFallbackContext } from './useExtensibleFallbackContext'
import { useTwapOrderCreationContext } from './useTwapOrderCreationContext'

import { DEFAULT_TWAP_EXECUTION_INFO } from '../const'
import { createTwapOrderTxs } from '../services/createTwapOrderTxs'
import { extensibleFallbackSetupTxs } from '../services/extensibleFallbackSetupTxs'
import { twapOrderAtom } from '../state/twapOrderAtom'
import { addTwapOrderToListAtom } from '../state/twapOrdersListAtom'
import { TwapOrderItem, TwapOrderStatus } from '../types'
import { buildTwapOrderParamsStruct } from '../utils/buildTwapOrderParamsStruct'
import { getConditionalOrderId } from '../utils/getConditionalOrderId'
import { twapOrderToStruct } from '../utils/twapOrderToStruct'

export function useCreateTwapOrder() {
  const { chainId, account } = useWalletInfo()
  const twapOrder = useAtomValue(twapOrderAtom)
  const addTwapOrderToList = useUpdateAtom(addTwapOrderToListAtom)

  const { inputCurrencyAmount, outputCurrencyAmount } = useAdvancedOrdersDerivedState()

  const appDataInfo = useAppData()
  const safeAppsSdk = useSafeAppsSdk()
  const twapOrderCreationContext = useTwapOrderCreationContext(inputCurrencyAmount as Nullish<CurrencyAmount<Token>>)
  const extensibleFallbackContext = useExtensibleFallbackContext()

  const uploadAppData = useUploadAppData()
  const tradeConfirmActions = useTradeConfirmActions()

  const { priceImpact } = useTradePriceImpact()
  const { confirmPriceImpactWithoutFee } = useConfirmPriceImpactWithoutFee()

  return useCallback(
    async (fallbackHandlerIsNotSet: boolean) => {
      if (!chainId || !account) return
      if (
        !inputCurrencyAmount ||
        !outputCurrencyAmount ||
        !twapOrderCreationContext ||
        !extensibleFallbackContext ||
        !safeAppsSdk ||
        !appDataInfo ||
        !twapOrder
      )
        return

      const isPriceImpactConfirmed = await confirmPriceImpactWithoutFee(priceImpact)

      if (!isPriceImpactConfirmed) {
        return
      }

      const pendingTrade = {
        inputAmount: inputCurrencyAmount,
        outputAmount: outputCurrencyAmount,
      }

      const twapFlowAnalyticsContext: SwapFlowAnalyticsContext = {
        account,
        recipient: twapOrder.receiver,
        recipientAddress: twapOrder.receiver,
        marketLabel: [inputCurrencyAmount.currency.symbol, outputCurrencyAmount.currency.symbol].join(','),
        orderClass: 'TWAP',
      }

      const paramsStruct = buildTwapOrderParamsStruct(chainId, twapOrder)
      const orderId = getConditionalOrderId(paramsStruct)

      tradeConfirmActions.onSign(pendingTrade)
      tradeFlowAnalytics.placeAdvancedOrder(twapFlowAnalyticsContext)
      twapConversionAnalytics('posted', fallbackHandlerIsNotSet)

      try {
        const fallbackSetupTxs = fallbackHandlerIsNotSet
          ? await extensibleFallbackSetupTxs(extensibleFallbackContext)
          : []
        const createOrderTxs = createTwapOrderTxs(twapOrder, paramsStruct, twapOrderCreationContext)
        const { safeTxHash } = await safeAppsSdk.txs.send({ txs: [...fallbackSetupTxs, ...createOrderTxs] })

        const orderItem: TwapOrderItem = {
          order: twapOrderToStruct(twapOrder),
          status: TwapOrderStatus.WaitSigning,
          chainId,
          safeAddress: account,
          submissionDate: new Date().toISOString(),
          id: orderId,
          executionInfo: DEFAULT_TWAP_EXECUTION_INFO,
        }

        addTwapOrderToList(orderItem)

        const summary = getOrderSubmitSummary({
          recipient: twapOrder.receiver,
          kind: OrderKind.SELL,
          recipientAddressOrName: null,
          account,
          inputAmount: twapOrder.sellAmount,
          outputAmount: twapOrder.buyAmount,
          feeAmount: undefined,
        })
        getCowSoundSend().play()
        dispatchPresignedOrderPosted(store, orderId, summary, OrderClass.LIMIT)

        uploadAppData({ chainId, orderId, appData: appDataInfo })
        tradeConfirmActions.onSuccess(safeTxHash)
        tradeFlowAnalytics.sign(twapFlowAnalyticsContext)
        twapConversionAnalytics('signed', fallbackHandlerIsNotSet)
      } catch (error) {
        tradeConfirmActions.onError(error.message || error)
        tradeFlowAnalytics.error(error, error.message, twapFlowAnalyticsContext)
        twapConversionAnalytics('rejected', fallbackHandlerIsNotSet)
      }
    },
    [
      chainId,
      account,
      inputCurrencyAmount,
      outputCurrencyAmount,
      twapOrder,
      tradeConfirmActions,
      twapOrderCreationContext,
      extensibleFallbackContext,
      addTwapOrderToList,
      safeAppsSdk,
      confirmPriceImpactWithoutFee,
      priceImpact,
      uploadAppData,
      appDataInfo,
    ]
  )
}
