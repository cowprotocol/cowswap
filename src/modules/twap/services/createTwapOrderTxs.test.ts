import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { CurrencyAmount } from '@uniswap/sdk-core'

import { COW } from 'legacy/constants/tokens'
import { WETH_GOERLI } from 'legacy/utils/goerli/constants'

import { COMPOSABLE_COW_ADDRESS, CURRENT_BLOCK_FACTORY_ADDRESS } from 'modules/advancedOrders'
import { getAppDataHash } from 'modules/appData'

import { createTwapOrderTxs } from './createTwapOrderTxs'

import { TwapOrderCreationContext } from '../hooks/useTwapOrderCreationContext'
import { TWAPOrder } from '../types'
import { buildTwapOrderParamsStruct } from '../utils/buildTwapOrderParamsStruct'

const chainId = SupportedChainId.GOERLI

const order: TWAPOrder = {
  sellAmount: CurrencyAmount.fromRawAmount(COW[SupportedChainId.GOERLI], 100_000_000_000),
  buyAmount: CurrencyAmount.fromRawAmount(WETH_GOERLI, 200_000),
  receiver: '0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6',
  numOfParts: 3,
  startTime: 1684764716,
  timeInterval: 600,
  span: 0,
  appData: getAppDataHash(),
}

const CREATE_COW_TX_DATA = '0xCREATE_COW_TX_DATA'
const APPROVE_TX_DATA = '0xAPPROVE_TX_DATA'

describe('Create TWAP order', () => {
  let context: TwapOrderCreationContext
  let createCowFn: jest.Mock
  let approveFn: jest.Mock

  beforeEach(() => {
    jest.spyOn(Date, 'now').mockImplementation(() => 1497076708000)

    createCowFn = jest.fn().mockReturnValue(CREATE_COW_TX_DATA)
    approveFn = jest.fn().mockReturnValue(APPROVE_TX_DATA)

    context = {
      composableCowContract: {
        interface: { encodeFunctionData: createCowFn },
        address: COMPOSABLE_COW_ADDRESS[chainId],
      } as any,
      needsApproval: false,
      spender: '0xB4FBF271143F4FBf7B91A5ded31805e42b222222',
      currentBlockFactoryAddress: CURRENT_BLOCK_FACTORY_ADDRESS[chainId],
      erc20Contract: { interface: { encodeFunctionData: approveFn } } as any,
    }
  })

  it('When sell token is approved, then should generate only creation transaction', () => {
    const paramsStruct = buildTwapOrderParamsStruct(chainId, order)
    const result = createTwapOrderTxs(order, paramsStruct, { ...context, needsApproval: false })

    expect(createCowFn).toHaveBeenCalledTimes(1)
    expect(createCowFn.mock.calls[0]).toMatchSnapshot()

    expect(result.length).toBe(1)
    expect(result).toMatchSnapshot()
  })

  it('When sell token is NOT approved, then should generate approval and creation transactions', () => {
    const paramsStruct = buildTwapOrderParamsStruct(chainId, order)
    const result = createTwapOrderTxs(order, paramsStruct, { ...context, needsApproval: true })

    expect(createCowFn).toHaveBeenCalledTimes(1)
    expect(createCowFn.mock.calls[0]).toMatchSnapshot()

    expect(approveFn).toHaveBeenCalledTimes(1)
    expect(approveFn.mock.calls[0]).toMatchSnapshot()

    expect(result.length).toBe(2)
    expect(result).toMatchSnapshot()
  })
})
