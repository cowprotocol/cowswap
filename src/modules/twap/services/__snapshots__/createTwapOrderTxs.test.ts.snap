// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 1`] = `
Array [
  "createWithContext",
  Array [
    Object {
      "handler": "0x910d00a310f7Dc5B29FE73458F47f519be547D3d",
      "salt": "0x00000000000000000000000000000000000000000000000000000015c90b9b2a",
      "staticInput": "0x00000000000000000000000091056d4a53e1faa1a84306d4deaec71085394bc8000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000000000000000000000000000000000007c2d24d55000000000000000000000000000000000000000000000000000000000001046a00000000000000000000000000000000000000000000000000000000646b782c000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000002580000000000000000000000000000000000000000000000000000000000000000bace6e636671f2df6e59ef1eb4e63dce6b646a70c51b004ae5a07a9fbe4356fc",
    },
    "0x0899c7DC280363d263Cc954506134F249CceC4a5",
    "0x",
    true,
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 2`] = `
Array [
  "approve",
  Array [
    "0xB4FBF271143F4FBf7B91A5ded31805e42b222222",
    "100000000000",
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 3`] = `
Array [
  "approve",
  Array [
    "0xB4FBF271143F4FBf7B91A5ded31805e42b222222",
    "0",
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 4`] = `
Array [
  Object {
    "data": "0xAPPROVE_TX_DATA",
    "operation": 0,
    "to": "0x91056D4A53E1faa1A84306D4deAEc71085394bC8",
    "value": "0",
  },
  Object {
    "data": "0xAPPROVE_TX_DATA",
    "operation": 0,
    "to": "0x91056D4A53E1faa1A84306D4deAEc71085394bC8",
    "value": "0",
  },
  Object {
    "data": "0xCREATE_COW_TX_DATA",
    "operation": 0,
    "to": "0xF487887DA5a4b4e3eC114FDAd97dc0F785d72738",
    "value": "0",
  },
]
`;

exports[`Create TWAP order When sell token is NOT approved, then should generate approval and creation transactions 1`] = `
Array [
  "createWithContext",
  Array [
    Object {
      "handler": "0x910d00a310f7Dc5B29FE73458F47f519be547D3d",
      "salt": "0x00000000000000000000000000000000000000000000000000000015c90b9b2a",
      "staticInput": "0x00000000000000000000000091056d4a53e1faa1a84306d4deaec71085394bc8000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000000000000000000000000000000000007c2d24d55000000000000000000000000000000000000000000000000000000000001046a00000000000000000000000000000000000000000000000000000000646b782c000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000002580000000000000000000000000000000000000000000000000000000000000000bace6e636671f2df6e59ef1eb4e63dce6b646a70c51b004ae5a07a9fbe4356fc",
    },
    "0x0899c7DC280363d263Cc954506134F249CceC4a5",
    "0x",
    true,
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved, then should generate approval and creation transactions 2`] = `
Array [
  "approve",
  Array [
    "0xB4FBF271143F4FBf7B91A5ded31805e42b222222",
    "100000000000",
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved, then should generate approval and creation transactions 3`] = `
Array [
  Object {
    "data": "0xAPPROVE_TX_DATA",
    "operation": 0,
    "to": "0x91056D4A53E1faa1A84306D4deAEc71085394bC8",
    "value": "0",
  },
  Object {
    "data": "0xCREATE_COW_TX_DATA",
    "operation": 0,
    "to": "0xF487887DA5a4b4e3eC114FDAd97dc0F785d72738",
    "value": "0",
  },
]
`;

exports[`Create TWAP order When sell token is approved, then should generate only creation transaction 1`] = `
Array [
  "createWithContext",
  Array [
    Object {
      "handler": "0x910d00a310f7Dc5B29FE73458F47f519be547D3d",
      "salt": "0x00000000000000000000000000000000000000000000000000000015c90b9b2a",
      "staticInput": "0x00000000000000000000000091056d4a53e1faa1a84306d4deaec71085394bc8000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000000000000000000000000000000000007c2d24d55000000000000000000000000000000000000000000000000000000000001046a00000000000000000000000000000000000000000000000000000000646b782c000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000002580000000000000000000000000000000000000000000000000000000000000000bace6e636671f2df6e59ef1eb4e63dce6b646a70c51b004ae5a07a9fbe4356fc",
    },
    "0x0899c7DC280363d263Cc954506134F249CceC4a5",
    "0x",
    true,
  ],
]
`;

exports[`Create TWAP order When sell token is approved, then should generate only creation transaction 2`] = `
Array [
  Object {
    "data": "0xCREATE_COW_TX_DATA",
    "operation": 0,
    "to": "0xF487887DA5a4b4e3eC114FDAd97dc0F785d72738",
    "value": "0",
  },
]
`;
